cmd/provisioner-localpv/app/helper.go

```diff
@@ -81,13 +87,90 @@ func (p *Provisioner) getPathAndNodeForPV(pv *v1.PersistentVolume) (path, node s
 	return path, node, nil
 }
 
+// createInitPod launches a helper(busybox) pod, to create the host path.
+//  The local pv expect the hostpath to be already present before mounting
+//  into pod. Validate that the local pv host path is not created under root.
+func (p *Provisioner) createInitPod(pOpts *HelperPodOptions) (err error) {
+	if pOpts.name == "" || pOpts.path == "" || pOpts.nodeName == "" {
+		return fmt.Errorf("invalid empty name or path or node")
+	}
+
+	// Initialize HostPath builder and validate that
+	// non-root directories are not passed for create
+	// Extract the base path and the volume unique path.
+	parentDir, volumeDir, vErr := hostpath.NewBuilder().WithPath(pOpts.path).
+		WithCheckf(hostpath.IsNonRoot(), "path should not be a root directory: path %v", pOpts.path).
+		ExtractSubPath()
+	if vErr != nil {
+		return vErr
+	}
+
+	conObj, _ := mContainer.Builder().
+		WithName("local-path-init").
+		WithImage(p.helperImage).
+		WithCommand(append(pOpts.cmdsForPath, filepath.Join("/data/", volumeDir))).
+		WithVolumeMounts([]v1.VolumeMount{
+			{
+				Name:      "data",
+				ReadOnly:  false,
+				MountPath: "/data/",
+			},
+		}).
+		Build()
+	containers := []v1.Container{conObj}
+
+	volObj, _ := mVolume.NewBuilder().
+		WithName("data").
+		WithHostDirectory(parentDir).
+		Build()
+	volumes := []v1.Volume{*volObj}
+
+	helperPod, _ := mPod.NewBuilder().
+		WithName("init-" + pOpts.name).
+		WithRestartPolicy(v1.RestartPolicyNever).
+		WithNodeName(pOpts.nodeName).
+		WithContainers(containers).
+		WithVolumes(volumes).
```

Comment:
Will it be more readable/intuitive if container & volume related builder are embedded inside pod builder?
Just image pod builder to accept a container builder as well as a volume builder.
```go
helperPod, _ := pod.NewBuilder().
  WithName("init-" + pOpts.name).
  WithRestartPolicy(v1.RestartPolicyNever).
  WithNodeName(pOpts.nodeName).
  WithContainer(
    container.NewBuilder().
      WithName(...).
      WithCommand(...).
      WithImage(...).
      WithVolumeMounts(...)
  ).
  WithVolume(
    volume.NewBuilder().
      WithName("data").
      WithHostDirectory(parentDir)
  ).
  Build()
```

Reply:
https://github.com/openebs/maya/issues/1231